overlord/snapstate/refresh.go

```diff
@@ -229,3 +230,21 @@ func softCheckNothingRunningForRefresh(st *state.State, snapst *SnapState, info
 		return inhibitRefresh(st, snapst, info, SoftNothingRunningRefreshCheck)
 	})
 }
+
+// SoftCheckNothingRunningForRefreshKeepLocked checks if non-service apps are off for a snap refresh and
+// keeps the snap lock on return. The lock must be released by the caller.
+func SoftCheckNothingRunningForRefreshKeepLocked(st *state.State, snapst *SnapState, info *snap.Info) (lock *osutil.FileLock, inhibited bool, err error) {
+	// Grab per-snap lock to prevent new processes from starting.
+	lock, err = snaplock.OpenLock(info.InstanceName())
```

Comment:
maybe my explanation was confusing but the lock we want to hold is the run inhibit lock, not the whole snap lock

Reply:
what we do here will interact with https://github.com/snapcore/snapd/pull/7700